### ISSUE TITLE
[BUG] Class based dependencies with __call__ don't work with yield

### ISSUE THREAD

**Issue by neriusmika** [tiangolo/fastapi#1204]
### Describe the bug

Class based callable dependencies (https://fastapi.tiangolo.com/advanced/advanced-dependencies/) don't work when used as generators with yield keyword.

### To Reproduce

```Python
class SessionScope:

    def __init__(self, uri):
        self.engine = create_engine(uri)
        self.session_maker = scoped_session(sessionmaker(bind=self.engine))

    def __call__(self):
        session = self.session_maker()
        try:
            yield session
        except:
            session.rollback()
            raise
        finally:
            self.session_maker.remove()

get_db = SessionScope('postgresql://postgres:mypass@127.0.0.1:5432/my_db')

@router.get('/', response_model=List[User])
def read_users(
    db: Session = Depends(get_db),
    skip: int = 0,
    limit: int = 100,
):
    """
    Retrieve users.
    """
    users = crud.user.get_multi(db, skip=skip, limit=limit)
    return users
```

### Expected behavior

Dependency must return yielded session, but now it returns \_\_call\_\_ generator instance

### Additional context

The problem lies in /fastapi/dependencies/utils.py file on line 498:
```Python
        elif inspect.isgeneratorfunction(call) or inspect.isasyncgenfunction(call):
            stack = request.scope.get("fastapi_astack")
            if stack is None:
                raise RuntimeError(
                    async_contextmanager_dependencies_error
                )  # pragma: no cover
            solved = await solve_generator(
                call=call, stack=stack, sub_values=sub_values
            )
```
Paramter _call_ passed to _inspect.isgeneratorfunction_ is instance of a class instead of it's \_\_call\_\_ method and this function always returns False.

**Comment by neriusmika** [comment]
Maybe you will find a better solution, but replacing all calls to _inspect.isgeneratorfunction_ and inspect.isasyncgenfunction_ to:
```Python
def is_generator_callable(call: Callable) -> bool:
    if hasattr(call, '__call__'):
        call = call.__call__
    return inspect.isgeneratorfunction(call)

def is_async_generator_callable(call: Callable) -> bool:
    if hasattr(call, '__call__'):
        call = call.__call__
    return inspect.isasyncgenfunction(call)
```
returns correct results.

**Comment by levchik** [comment]
Got bitten by this one as well, @neriusmika would you mind making a PR with suggested changes?

**Comment by dhkron** [comment]
This one got me too..

**Comment by tiangolo** [comment]
Thanks for the discussion here everyone! :coffee: 

This was solved in https://github.com/tiangolo/fastapi/pull/1365 by @mrosales :heavy_check_mark: :tada: 

It will be available in version `0.56.1` (in a couple of hours).

**Comment by fanck0605** [comment]
@nmrtv @tiangolo 

Hello, may I ask if there are any specific considerations for replacing `asyncio.iscoroutinefunction` with `inspect.iscoroutinefunction`?

In Python version 3.9, this change will prevent asynchronous functions compiled with Cython from executing correctly, as `inspect.iscoroutinefunction(cythonized_async_function)` will return `False`. https://github.com/cython/cython/issues/2273#issuecomment-1188664872

Is it possible to revert this method back to asyncio.iscoroutinefunction? If possible, I will submit a pull request.
